File: packages/core/src/ide/workspaceInstructions.ts
/**
 * Copyright 2026 CodeSetu Contributors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { parseFrontmatter, splitFrontmatter } from "./frontmatter.js";
import type {
  WorkspaceInstruction,
  WorkspaceInstructionParseResult,
  WorkspaceInstructionSource,
} from "./types.js";

const REQUIRED_FIELDS = ["id", "name", "description"] as const;

export function parseWorkspaceInstructions(
  sources: WorkspaceInstructionSource[],
): WorkspaceInstructionParseResult {
  const result: WorkspaceInstructionParseResult = {
    skills: [],
    checks: [],
    warnings: [],
  };
  const seenIds = new Set<string>();

  for (const source of sources) {
    const parsed = parseSource(source);

    if (parsed.warning !== undefined) {
      result.warnings.push(parsed.warning);
      continue;
    }

    const instruction = parsed.instruction;

    if (seenIds.has(instruction.id)) {
      result.warnings.push(`${source.path}: duplicate instruction id "${instruction.id}"`);
      continue;
    }

    seenIds.add(instruction.id);

    if (instruction.kind === "skill") {
      result.skills.push(instruction);
    } else {
      result.checks.push(instruction);
    }
  }

  return result;
}

function parseSource(
  source: WorkspaceInstructionSource,
):
  | { instruction: WorkspaceInstruction; warning?: never }
  | { instruction?: never; warning: string } {
  const split = splitFrontmatter(source.content);

  if (split === undefined) {
    return { warning: `${source.path}: missing YAML frontmatter` };
  }

  const { body } = split;
  const fields = parseFrontmatter(split.frontmatter).scalars;

  for (const field of REQUIRED_FIELDS) {
    const value = fields[field];

    if (value === undefined || value.length === 0) {
      return { warning: `${source.path}: missing required field "${field}"` };
    }
  }

  if (body.length === 0) {
    return { warning: `${source.path}: empty instruction body` };
  }

  const { id, name, description } = fields;

  if (id === undefined || name === undefined || description === undefined) {
    return { warning: `${source.path}: missing required frontmatter` };
  }

  return {
    instruction: {
      kind: source.kind,
      path: source.path,
      id,
      name,
      description,
      body,
    },
  };
}

File: apps/jetbrains/src/main/kotlin/ai/codesetu/instructions/WorkspaceInstructions.kt
package ai.codesetu.instructions

import ai.codesetu.model.WorkspaceInstruction

data class WorkspaceInstructionSource(
  val kind: String,
  val path: String,
  val content: String,
)

data class WorkspaceInstructionParseResult(
  val skills: List<WorkspaceInstruction>,
  val checks: List<WorkspaceInstruction>,
  val warnings: List<String>,
)

fun parseWorkspaceInstructions(
  sources: List<WorkspaceInstructionSource>,
): WorkspaceInstructionParseResult {
  val skills = mutableListOf<WorkspaceInstruction>()
  val checks = mutableListOf<WorkspaceInstruction>()
  val warnings = mutableListOf<String>()
  val seenIds = mutableSetOf<String>()

  for (source in sources) {
    val parsed = parseOne(source)
    val warning = parsed.warning

    if (warning != null) {
      warnings.add(warning)
      continue
    }

    val instruction = parsed.instruction ?: continue
    if (!seenIds.add(instruction.id)) {
      warnings.add("${source.path}: duplicate instruction id \"${instruction.id}\"")
      continue
    }

    if (source.kind == "skill") {
      skills.add(instruction)
    } else {
      checks.add(instruction)
    }
  }

  return WorkspaceInstructionParseResult(skills, checks, warnings)
}

private data class ParsedInstruction(
  val instruction: WorkspaceInstruction? = null,
  val warning: String? = null,
)

private fun parseOne(source: WorkspaceInstructionSource): ParsedInstruction {
  val regex = Regex("^---\\n([\\s\\S]*?)\\n---\\n?([\\s\\S]*)$")
  val match = regex.find(source.content)
    ?: return ParsedInstruction(warning = "${source.path}: missing YAML frontmatter")

  val metadata = parseSimpleYaml(match.groupValues[1])
  val id = metadata["id"]?.trim().orEmpty()
  val name = metadata["name"]?.trim().orEmpty()
  val description = metadata["description"]?.trim().orEmpty()
  val body = match.groupValues[2].trim()

  if (id.isEmpty() || name.isEmpty() || description.isEmpty()) {
    return ParsedInstruction(warning = "${source.path}: id, name, and description are required")
  }

  if (body.isEmpty()) {
    return ParsedInstruction(warning = "${source.path}: instruction body is required")
  }

  return ParsedInstruction(
    instruction = WorkspaceInstruction(
      id = id,
      name = name,
      description = description,
      sourcePath = source.path,
      body = body,
    ),
  )
}

private fun parseSimpleYaml(yaml: String): Map<String, String> =
  yaml.lines()
    .mapNotNull { line ->
      val separator = line.indexOf(":")
      if (separator == -1) {
        null
      } else {
        line.take(separator).trim() to line.drop(separator + 1).trim().trim('"', '\'')
      }
    }
    .toMap()
